### PR TITLE
added STL datatype

### PR DESCRIPTION
**Pull request by ThomasWollmann**

### CODE REVIEW

**Comment by ThomasWollmann** [comment]
Added image support to IEs as well. Do you mind if I don't open a second pull request for this minor change?

File: config/datatypes_conf.xml.sample [hunk 1]
```diff
@@ -436,10 +436,11 @@
     <datatype extension="oxliss" type="galaxy.datatypes.binary:OxliSubset" mimetype="application/octet-stream" display_in_upload="true"/>
     <datatype extension="oxligl" type="galaxy.datatypes.binary:OxliGraphLabels" mimetype="application/octet-stream" display_in_upload="true"/>
     <!-- Constructive solid geometry datatypes -->
-    <datatype extension="plyascii" type="galaxy.datatypes.constructive_solid_geometry:PlyAscii" display_in_upload="true" />
-    <datatype extension="plybinary" type="galaxy.datatypes.constructive_solid_geometry:PlyBinary" display_in_upload="true" />
-    <datatype extension="vtkascii" type="galaxy.datatypes.constructive_solid_geometry:VtkAscii" display_in_upload="true" />
-    <datatype extension="vtkbinary" type="galaxy.datatypes.constructive_solid_geometry:VtkBinary" display_in_upload="true" />
+    <datatype extension="stl" type="galaxy.datatypes.constructive_solid_geometry:STL" display_in_upload="True"/>
```
Comment (bgruening) [inline config/datatypes_conf.xml.sample]: s/True/true just for being consistent.

File: lib/galaxy/dependencies/pinned-requirements.txt [hunk 2]
```diff
@@ -69,3 +69,5 @@ ecdsa==0.13
 # Flexible BAM index naming
 pysam==0.8.4+gx1
 
+# CSG dependencies
+numpy-stl==1.8.0
```
Comment (bgruening) [inline lib/galaxy/dependencies/pinned-requirements.txt]: Thomas for generating a wheel that is automatically shipped with Galaxy you need to create a wheel on our build-server with https://github.com/galaxyproject/starforge

Example PR is for example here: https://github.com/galaxyproject/starforge/pull/100

File: lib/galaxy/datatypes/constructive_solid_geometry.py [hunk 3]
```diff
@@ -110,7 +111,7 @@ def display_peek(self, dataset):
 
 class PlyAscii(Ply, data.Text):
 
-    file_ext = "plyascii"
+    file_ext = "ply"
```
Comment (bgruening) [inline lib/galaxy/datatypes/constructive_solid_geometry.py]: @gregvonkuster do you see any negative impact in changing this?

Comment (nsoranzo) [inline lib/galaxy/datatypes/constructive_solid_geometry.py]: Yes, this and the following changes are wrongly assigning the same value to the `file_ext` of different classes: this attribute should be unique and equal to what is used as `extension` in `config/datatypes_conf.xml.sample` .

Comment (gregvonkuster) [inline lib/galaxy/datatypes/constructive_solid_geometry.py]: @nsoranzo is correct - there are binary and ascii types of the ply datatype, so we need to differentiate them.

Comment (ThomasWollmann) [inline lib/galaxy/datatypes/constructive_solid_geometry.py]: Still many tools don't distinguish between the extensions. Havn't seen any tool which saves a file in xxx.vtkascii. A correct file extension is mandatory for https://github.com/ThomasWollmann/docker-paraviewweb.

Comment (bgruening) [inline lib/galaxy/datatypes/constructive_solid_geometry.py]: Thomas, for the time being can you leave the extensions as they are and use in paraview a if condition like

```
#if $hda.ext in ['plybinary', 'plyascii']:
   cp $hda foo.ply 
#end 
```

Comment (ThomasWollmann) [inline lib/galaxy/datatypes/constructive_solid_geometry.py]: Very dirty, but I'll do.

**Comment by nsoranzo** [comment]
@galaxybot test this

File: lib/galaxy/dependencies/pinned-requirements.txt (hunk 2 above)
Comment (bgruening) [inline lib/galaxy/dependencies/pinned-requirements.txt]: We should have now a wheel for version 1.9.1
https://github.com/galaxyproject/starforge/pull/104
Can you adjust this and then we are ready to go.

**Comment by bgruening** [comment]
@ThomasWollmann the good point is we are now using the numpy-stl wheel and you can rely on it. The bad part is that your sniffer is not specific enough. Other files are also recognised as stl: https://travis-ci.org/galaxyproject/galaxy/jobs/149693586#L1166

**Comment by ThomasWollmann** [comment]
Since it is not trivial for the binary STLs I used the library. Sadly it is not a working solution. I'll think about a new solution.

**Comment by bgruening** [comment]
Maybe it's enough to check actually the content of your mesh. I assume that reading a text file or an other hdf5 store simply do not raise an exception, but the content is probably empty.

File: lib/galaxy/datatypes/constructive_solid_geometry.py [hunk 4]
```diff
@@ -8,6 +8,9 @@
 from galaxy.datatypes.metadata import MetadataElement
 from galaxy import util
 
+from stl import mesh
+import numpy as np
+
```
Comment (nsoranzo) [inline lib/galaxy/datatypes/constructive_solid_geometry.py:12]: Can you invert this 2 imports and move them before the `galaxy` imports? I.e.:

```
import numpy as np
from stl import mesh

from galaxy.datatypes import data
...
```

We are now following the `smarkets` import style, as in this [example](https://github.com/PyCQA/flake8-import-order/blob/master/tests/test_cases/complete_smarkets.py).

**Comment by jmchilton** [comment]
I'm going to mark this as WIP until the above comments are resolved. Thanks for the initial crack at it and good luck improving the sniffer.

**Comment by ThomasWollmann** [comment]
Still unsure if this is possible. Should I remove the sniffer so we can use the datatypes without automatic detection?

**Comment by bgruening** [comment]
Yes, let's remove the sniffer.

**Comment by bgruening** [comment]
Looks good to me. Hopefully we will find a way to sniff this in the future. 
@nsoranzo what do you think can we merge this?

**Comment by bgruening** [comment]
Thanks @ThomasWollmann!

**Pull request merged**